### PR TITLE
Feature/cdar 335 sensorlib xmlrpc interface

### PR DESCRIPTION
**Pull request by davidgayman**
# PR Details
## Description

Added XML-RPC server support. Object detection is executed in a loop on a separate thread, and RPC call retrieves cached objects. Changed serialization to string replacing list-of-strings and updated unit test.

## Related Jira Key

CDAR-335

## Motivation and Context

Sensorlib capabilities need to be accessible in a Java context, the RPC server provides this functionality through an exposed function.

## How Has This Been Tested?

Local testing has confirmed correct initial function calls, but did not validate with a running CARLA instance. Unit tests were reviewed and passing.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by paulbourelly999** [comment]
@davidgayman Please in the future avoid making single PRs this large. After skimming the material in this PR it looks like there are several parallel changes in it. In the future please put single PR up for each. This will allow individual PRs to be done quicker and reviewers to be able to focus more on the content of changes. 
I know when working on code it can sometimes be easy to lump in fixes you discover into a single PR but this normally makes the PR difficult to review. Just mentioning this as a future improvement, so for this PR its fine but any future bugs or necessary enhancements please standup independent smaller PRs where possible

File: sensorlib/ReadMe.md [hunk 1]
```diff
@@ -12,6 +12,14 @@ Raw sensor observations from the CARLA sensor are composed to build a dataset re
 
 This new feature is provided to enable vulnerable road user detection using a LIDAR sensor in the context of a CARLA-driven simulation. The data is intended to feed into a CARMA V2X system for data fusion with other sensors and action by CARMA platform vehicles.
 
+## Interface
+
+The XML-RPC interface consists of one function to retrieve the list of detected objects.
```
Comment (paulbourelly999) [inline sensorlib/ReadMe.md:17]: There should be 2 functions:
- One to create a sensor at a given location/orientation.
- One to retrieve all detected objects from the sensor.

File: sensorlib/src/main.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    :return: No return if RPC server is running as this blocks execution; The SimulatedLidarSensor object if RPC server
+                is not running.
+    """
+    # Get inputs
+    sensor_transform = carla.Transform(carla.Location(x=0.0, y=0.0, z=0.0),
+                                       carla.Rotation(pitch=0.0, yaw=0.0, roll=0.0))
+
+    # Retrieve the sensor configuration
+    if os.path.exists(sensor_config):
+        config = SimulatedSensorUtils.load_config_from_file(sensor_config)
```
Comment (paulbourelly999) [inline sensorlib/src/main.py:67]: We need to make a separate PR after this to correct the current API. The sensor library should not spawn any sensors on startup and instead should be able to dynamically spawn sensors based on requests sent by the CARLA Ambassador in CDASim.

File: sensorlib/test/test_SemanticLidarSensor.py [hunk 3]
```diff
@@ -125,6 +125,9 @@ def test_get_detected_objects_json(self):
         with open("data/test_data_serialized_detected_objects.json", "r") as file:
             expected_serialized_data = json.load(file)
             assert serialized == expected_serialized_data
+        # with open("data/test_data_serialized_detected_objects.json", "w") as file:
```
Comment (paulbourelly999) [inline sensorlib/test/test_SemanticLidarSensor.py:128]: Why is this commented out.

File: sensorlib/src/main.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    scheduler.enter(detection_cycle_delay_seconds, 1, scheduled_compute, (scheduler, simulated_lidar_sensor,
+                                                                          detection_cycle_delay_seconds))
+    scheduler_thread = threading.Thread(target=scheduler.run)
+    print("Starting sensorlib compute.")
+    scheduler_thread.start()
+
+    # Create an XML-RPC server
+    if start_rpc_server:
+        print("Starting sensorlib XML-RPC server.")
+        server = SimpleXMLRPCServer((xmlrpc_server_host, xmlrpc_server_port))
```
Comment (EricChen-Lei) [inline sensorlib/src/main.py:103]: I think you might need to init a xmlrpc.server.SimpleXMLRPCRequestHandler, since this deals with the POST request

Comment (dg-developer) [reply sensorlib/src/main.py:110]: Looks like this is automatic. Default parameter.

File: sensorlib/src/main.py (hunk 2 above)
Comment (dg-developer) [reply sensorlib/src/main.py:72]: That makes sense. One question: Are we hoping to run the sensorlib inside some other code such as the integration tool or as a standalone server?

File: sensorlib/ReadMe.md (hunk 1 above)
Comment (dg-developer) [reply sensorlib/ReadMe.md:17]: Are we only exposing as a server? If so then yes. I was under the impression some other Python code would configure the sensor and then we serve the detected objects, but if this needs to be a standalone server you are correct.

File: sensorlib/test/data/test_data_serialized_detected_objects.json [hunk 5]
```diff
@@ -0,0 +1 @@
...[100.0, 1.0, 0.0], \\\"rotation\\\": [[0.9983314336437902, -0.02072298021475338, 0.05389718628791056], [0.02439869428620028, 0.9973554635114954, -0.06846008414323508], [-0.05233595624294384, 0.06966087492121549, 0.9961969233988568]], \\\"angular_velocity\\\": [0.0, 0.0, 8.726646064915904e-05], \\\"position_covariance\\\": [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]], \\\"velocity_covariance\\\": [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]], \\\"confidence\\\": 1.0}\", \"{\\\"carla_actor\\\": null, \\\"id\\\": 5, \\\"object_type\\\": \\\"Vehicles\\\", \\\"timestamp\\\": 0, \\\"bounding_box_in_world_coordinate_frame\\\": [], \\\"position\\\": [20.0, 34.64101791381836, 0.0], \\\"velocity\\\": [100.0, 1.0, 0.0], \\\"rotation\\\": [[0.9983314336437902, -0.02072298021475338, 0.05389718628791056], [0.02439869428620028, 0.9973554635114954, -0.06846008414323508], [-0.05233595624294384, 0.06966087492121549, 0.9961969233988568]], \\\"angular_velocity\\\": [0.0, 0.0, 8.726646064915904e-05], \\\"position_covariance\\\": [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]], \\\"velocity_covariance\\\": [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]], \\\"confidence\\\": 1.0}\"]"
```
[the remark this answers is not in the document]
Comment (dg-developer) [reply sensorlib/test/data/test_data_serialized_detected_objects.json:1]: As best as I can figure out how to serialize a list of objects into a single JSON string, this needs to be performed in two steps: Serialize each object, then serialize the collection. For consumption by the Python json library the result likely needs to respect this double-serialization. However we could consider adding a cleanup step before sending to external systems. This needs to be discussed if we want to understand the cleanest approach.

File: sensorlib/test/test_SemanticLidarSensor.py (hunk 3 above)
Comment (dg-developer) [reply sensorlib/test/test_SemanticLidarSensor.py:128]: Debugging. Needs to be removed.

File: sensorlib/src/SimulatedSensorConfigurator.py [hunk 6]
```diff
@@ -81,13 +90,12 @@ def get_simulated_sensor(infrastructure_id):
...
     @staticmethod
-    def __get_initialized_carla_world(simulated_sensor_config):
+    def __get_initialized_carla_world(carla_host, carla_port):
         """Initialize the CARLA connection, which only needs to be done once."""
 
         if SimulatedSensorConfigurator.__client is None:
-            SimulatedSensorConfigurator.__client = carla.Client(
-                simulated_sensor_config["carla_connection"]["carla_host"],
-                simulated_sensor_config["carla_connection"]["carla_port"])
+            SimulatedSensorConfigurator.__client = carla.Client("localhost", 2000)
```
Comment (paulbourelly999) [inline sensorlib/src/SimulatedSensorConfigurator.py:97]: Overwriting configured values

File: sensorlib/test/data/test_data_serialized_detected_objects.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (EricChen-Lei) [reply sensorlib/test/data/test_data_serialized_detected_objects.json:1]: Just curious, for the rotation, why are there 3 sets of arrays, I thought it should just be [x y z w]

**Comment by davidgayman** [comment]
Unit tests are currently failing.

File: sensorlib/ReadMe.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+## Overview
+
+The sensor library provides a proxy sensor for the CARLA simulator, exposing additional features.
+
+This new feature is provided to enable vulnerable road user detection using a LIDAR sensor in the context of a
+CARLA-driven simulation. The data is intended to feed into a CARMA V2X system for data fusion with other sensors and
+action by CARMA platform vehicles.
+
+The SimulatedSensor acts as a wrapper to the CARLA Semantic Lidar Sensor, capturing the data and performing
```
Comment (paulbourelly999) [inline sensorlib/ReadMe.md:11]: Can we add information about why we are wrapping the CARLA Semantic Lidar?

File: sensorlib/ReadMe.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+post-processing to generate a list of detected objects upon request.
+
+Raw sensor observations from the CARLA sensor are analyzed to build a dataset representing the current objects in scene.
+
+## Service Layer
```
Comment (paulbourelly999) [inline sensorlib/ReadMe.md:16]: Please rename this service to CARLA CDASim Adapter. This CARLA library gets used here and in Eric's CARLA CARMA Integration tool so there isn't really one service where this gets used.

File: sensorlib/.gitignore [hunk 9]
```diff
@@ -1,3 +1,4 @@
 __pycache__/
 .project
 .pydevproject
+machinet.conf
```
Comment (paulbourelly999) [inline sensorlib/.gitignore:4]: Is this some IDE plugin configuration file?

File: sensorlib/config/noise_model_config.yaml [hunk 10]
```diff
@@ -1,4 +1,9 @@
 noise_model_name: GaussianNoiseModel
+stages:
+    position_noise: True
+    orientation_noise: True
+    type_noise: True
```
Comment (paulbourelly999) [inline sensorlib/config/noise_model_config.yaml:5]: We talked about defaulting to turn of type noise since its implementation is incomplete currently right. Last we talked I remember this noise essentially assigning a random type classification for each detection which we concluded is not really noise and more of just invalid data. Noise would be introducing some error to an otherwise valid signal. Making the entire signal random error is not really noise. Let me know if anything has changed.

File: sensorlib/test/data/test_data_serialized_detected_objects.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply sensorlib/test/data/test_data_serialized_detected_objects.json:1]: @EricChen-Lei I think this json includes multiple elements, each one has a rotation array of x,y,z

File: sensorlib/src/SimulatedSensorConfigurator.py (hunk 6 above)
Comment (davidgayman) [reply sensorlib/src/SimulatedSensorConfigurator.py]: Fixed

**Comment by davidgayman** [comment]
> Unit tests are currently failing.

Currently passing.

File: sensorlib/ReadMe.md (hunk 7 above)
Comment (davidgayman) [reply sensorlib/ReadMe.md:11]: Sure how's this:

`The SimulatedSensor acts as a wrapper to the CARLA Semantic Lidar Sensor, capturing the data and performing
post-processing to generate a list of detected objects upon request. The primary reason for processing data from CARLA
is to ensure noise is applied to the output signal, and customization is possible for occlusion and any other procedures
in the future. The native CARLA sensor does not provide signal noise, misdetection simulation, or customization.

`

File: sensorlib/ReadMe.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+The SimulatedSensor manages a CarlaSensor, which is a light wrapper around the CARLA sensor reference.
+
+High-speed LIDAR data is collected into a DataCollector class running in a separate context which self-populates via a
+callback.
+
+The SimulatedSensor.compute_detected_objects() function exposes the primary processing sequence performing the
+following stages:
+
+1. Truth state retrieval.
+2. Pre-filter by distance from sensor, and allowed object types.
```
[the remark this answers is not in the document]
Comment (davidgayman) [reply sensorlib/ReadMe.md:37]: Referring to this parameter?

`range | float | 10.0 | Maximum distance to measure/raycast in meters (centimeters for CARLA 0.9.6 or previous).
-- | -- | -- | --`

https://carla.readthedocs.io/en/latest/ref_sensors/#semantic-lidar-sensor

File: sensorlib/config/noise_model_config.yaml (hunk 10 above)
Comment (davidgayman) [reply sensorlib/config/noise_model_config.yaml:5]: False by default now. This is a hook for future robust implementation.

File: sensorlib/ReadMe.md (hunk 7 above)
Comment (paulbourelly999) [reply sensorlib/ReadMe.md:11]: yeah that is fine.

File: sensorlib/ReadMe.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply sensorlib/ReadMe.md:37]: Yeah

File: sensorlib/ReadMe.md [hunk 12]
```diff
@@ -0,0 +1,88 @@
+# CARLA Sensor Library
+
+## Overview
+
+The sensor library provides a proxy sensor for the CARLA simulator, exposing additional features.
+
+This new feature is provided to enable vulnerable road user detection using a LIDAR sensor in the context of a
```
Comment (paulbourelly999) [inline sensorlib/ReadMe.md:7]: This sentence is probably slightly inaccurate? The feature is to enable object detection for our CDASim deployment using CARLA's existing sensor and object modeling ("object modeling" might not be the right word here either. What I am referring is CARLAs 3d world model that include dynamic vehicles and pedestrians and static obstacles).

Comment (paulbourelly999) [reply sensorlib/ReadMe.md:7]: The main concern here is that this is not a CARLA-Driven simulation, we are just using CARLA for vehicle dynamics and now for 3d world modeling and object detection.

File: sensorlib/ReadMe.md [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+CARLA-driven simulation. The data is intended to feed into a CARMA V2X system for data fusion with other sensors and
+action by CARMA platform vehicles.
```
Comment (paulbourelly999) [inline sensorlib/ReadMe.md:9]: This data will be feed to CARMA-Platform via the CARLA CARMA Integration tool and to CDASim for any other integrated simulators or software systems under test that need it.

File: sensorlib/ReadMe.md [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+The SimulatedSensor acts as a wrapper to the CARLA Semantic Lidar Sensor, capturing the data and performing
+post-processing to generate a list of detected objects upon request. The primary reason for processing data from CARLA
+is to ensure noise is applied to the output signal, and customization is possible for occlusion and any other procedures
+in the future. The native CARLA sensor does not provide signal noise, misdetection simulation, or customization.
```
Comment (paulbourelly999) [inline sensorlib/ReadMe.md:14]: Lets include some information about the fact that we are currently also fixing a CARLA 0.9.10 bug with actor ID accuracy for returned data.

File: sensorlib/ReadMe.md [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+Raw sensor observations from the CARLA sensor are analyzed to build a dataset representing the current objects in scene.
+
+## Carla CDA Sim Adapter
+
+The Carla CDA Sim Adapter exposes the sensing capability through an RPC service. See
+the [Carla CDA Sim Adapter Documentation](doc/CarlaCDASimAdapter.md) for details.
+
+## Carla CDA Sim API
```
Comment (paulbourelly999) [inline sensorlib/ReadMe.md:23]: This is not what I meant. I meant rename the XML RPC service to CARLA CDA Sim Adapter. The library should still be called  what ever we are calling the sensor library. I assume this is the API you are referring to here.

File: sensorlib/_requirements.txt [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+cryptography
+idna
+mkl-fft==1.3.6
+mkl-random
+mkl-service==2.4.0
+numpy
+packaging
+pooch
+pycparser
+pyOpenSSL
```
Comment (paulbourelly999) [inline sensorlib/_requirements.txt:15]: What are some of these dependencies for?

File: sensorlib/_requirements.txt [hunk 17]
```diff
@@ -0,0 +1,20 @@
+appdirs==1.4.4
+brotlipy==0.7.0
+certifi
+cffi
+charset-normalizer
+cryptography
```
Comment (paulbourelly999) [inline sensorlib/_requirements.txt:6]: what is this package for?

File: sensorlib/_requirements.txt [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+numpy
+packaging
+pooch
+pycparser
+pyOpenSSL
+PySocks
+PyYAML
+requests
+scipy==1.10.1
+urllib3
```
Comment (paulbourelly999) [inline sensorlib/_requirements.txt:20]: are these requirements of the python CARLA client

File: sensorlib/doc/CarlaCDASimAPI.md [hunk 19]
```diff
@@ -0,0 +1,208 @@
+# Carla CDA Sim API
```
Comment (paulbourelly999) [inline sensorlib/doc/CarlaCDASimAPI.md:1]: Same comment about the naming here, we can revert this to Sensor Library API or what ever you had initially called this since it is not intrinsically connect to CDA Sim at all.

File: sensorlib/doc/CarlaCDASimAPI.md [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+api = CarlaCDASimAPI.build_from_client(carla_client)
+
+# or...
+
+api = CarlaCDASimAPI.build_from_world(carla_world)
+```
+
+The API object may then be used to indirectly build and manage sensors, and request output data. Sensors are located by
+association with an infrastructure item, such as a traffic light post, and the sensor index within a co-located
+collection of sensors. For example, a sensor with an infrastructure_id of 3 and sensor_id 0, is the 0th sensor
```
Comment (paulbourelly999) [inline sensorlib/doc/CarlaCDASimAPI.md:32]: Lets create a future issue to rename `infrastructure_id` to something else. It conflicts with CDASim's definition of infrastructure_id. Not terribly important but we want some thing to track this. Once we move the code somewhere. I will try to add a github issue to address this.

File: sensorlib/doc/CarlaCDASimAPI.md [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+```
+
+Objects can be retrieved from the sensor directly:
+
+```
+detected_objects = sensor.get_detected_objects()
+```
+
+Objects can also be retrieved from another function context by using the API. (Note: Currently sensor access from the
+API is only supported within the running Python context.)
```
Comment (paulbourelly999) [inline sensorlib/doc/CarlaCDASimAPI.md:64]: What distinction are you describing here.

File: sensorlib/requirements.txt [hunk 22]
```diff
@@ -0,0 +1,21 @@
+appdirs==1.4.4
+brotlipy==0.7.0
+carla==0.9.14
```
[the remark this answers is not in the document]
Comment (davidgayman) [reply sensorlib/requirements.txt]: Yes exactly.

File: sensorlib/doc/CarlaCDASimAPI.md [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+
+The Carla CDA Sim API provides programmatic access to the library functionality. The primary means of
+accessing features is through a CarlaCDASimAPI object, which may be constructed from any piece of information sufficient to
+establish a Carla connection.
+
+First ensure the following environment variables are set which define the CARLA version being utilized:
+
+| Environment Variable | Description                                                                                                       |
+| -------------------- |-------------------------------------------------------------------------------------------------------------------|
+| LOAD_CARLA_EGG       | Set to True to load the CARLA Python API from an egg file. Leave blank to use a pip-managed CARLA client library. |
```
[the remark this answers is not in the document]
Comment (davidgayman) [reply sensorlib/doc/CarlaCDASimAPI.md:11]: Seems like we ultimately want this library to support versions 9.10 through 9.14. We have to accommodate any differences, which are fortunately very minor. The alternative would be to make a version-specific library.

File: sensorlib/doc/CarlaCDASimAPI.md (hunk 20 above)
Comment (davidgayman) [reply sensorlib/doc/CarlaCDASimAPI.md:32]: That's understandable if there is a difference between this concept and cda infrastructure ID. Are they not the same?

File: sensorlib/doc/CarlaCDASimAPI.md (hunk 21 above)
Comment (davidgayman) [reply sensorlib/doc/CarlaCDASimAPI.md:64]: Context.
